### DISCUSSION TITLE
Nesting Observations in Days?

### DISCUSSION THREAD

**Discussion by carlotta-hi** [cdriveraus/ctsem#44] [Q&A]
Hi Charles,

thanks for all the support you've given me so far! 
I would like to use ctsem for ESM data analysis. The dataset contains 4 observations per day over 4 weeks and I am wondering if I need to nest the observations in days as one would do with traditional multilevel models?

Best
Carlotta

**Comment by cdriveraus** [comment]
Well, whether you need to nest depends on things like research question and the relevance of within day structure to your data, whether it's a regression model or a ct model. More than 2 levels of random effects gets tricky, but this kind of dependence can also be addressed through use of models with periodicity and or time dependent predictor effects.

**Reply by rifatzahan** [reply]
Is there any example in any documentation if someone wants to assess the reciprocal relationship between two variables, which were obtained multiple times daily for several days?

**Reply by cdriveraus** [reply]
The examples in the help for ctStanFit show a bivariate model, or here's an example from my teaching. If there's something about the day structure you want to include then you need to be more specific, with just a few days I doubt you want to do anything though. 

```

library(ctsem)

#generating data - can ignore
set.seed(3)
gm <- ctModel(LAMBDA=diag(2),MANIFESTVAR = diag(.5,2),
  DIFFUSION=matrix(c(.8,.5,0,.2),2,2),DRIFT=matrix(c(-.5,0,0,-.3),2,2),
  T0MEANS = c(20,-10),
  MANIFESTTRAITVAR=matrix(c(1,.7,0,.2),2,2),
  # MANIFESTMEANS=c(x,x*.4+rnorm(1,0,.6)),
  Tpoints=10,manifestNames = c('Stress','Quality'))
dat <-data.frame(ctGenerate(ctmodelobj = gm,n.subjects = 100,burnin = 2,wide = F,dtmean=.5,logdtsd = .5))

head(dat)


#model specification -- intercepts and initial parameters default to indivdually varying, 
# but we specify here in any case as a demonstration (using the | separator notation) 

model <- ctModel(
  LAMBDA = diag(2),
  type = 'stanct', #could also choose standt for discrete time (ie difference equation) modelling.
  manifestNames = c('Stress','Quality'),
  latentNames = c('Stress','Quality'),
  T0MEANS = c('t0_stress','t0_quality'), #starting values for latent processes
  MANIFESTMEANS = c('mi_stress||TRUE','mi_quality||TRUE'), #intercepts on observed variables
  MANIFESTVAR = c( #vectors are coerced rowwise into matrices of the right shape
    'merr_stress',0,
    0,'merr_quality'),
  DRIFT = c(  #temporal dependence matrix
    'a11','a12',
    'a21','a22'),
  DIFFUSION = c( #system noise. sd's on diagonal, unconstrained correlations on lower triangle
    'diff11',0,
    'diff21','diff22'),
  CINT = 0 #single values are repeated as many times as necessary (here, twice, one for each process)
  )

ctModelLatex(model) #view model equations, requires latex compiler

fit <- ctStanFit(datalong = dat,ctstanmodel = model, cores=2)

summary(fit)

ctModelLatex(fit) #view dynamic system estimates in equation form, requires latex compiler

ctKalman(fit,kalmanvec=c('y','yprior'),plot=TRUE, subjects=1:2) #forward predictions
ctKalman(fit,kalmanvec=c('y','ysmooth'),plot=TRUE, subjects=1:2) #best estimates conditional on all data
ctKalman(fit,kalmanvec=c('y','yprior'),plot=TRUE, subjects=1,removeObs = T) #predictions before seeing any data

ctStanDiscretePars(fit,plot=T)




#what if we disable individual differences in the intercepts, i.e. assume everyone has the same average level?

badmodel <- ctModel(LAMBDA = diag(2),
  type = 'stanct', #could also choose standt for discrete time (ie difference equation) modelling.
  manifestNames = c('Stress','Quality'),
  latentNames = c('Stress','Quality'),
  T0MEANS = c('t0_stress','t0_quality'), #starting values for latent processes
  MANIFESTMEANS = c('mi_stress||FALSE','mi_quality||FALSE'), #intercepts on observed variables
  MANIFESTVAR = c( #vectors are coerced rowwise into matrices of the right shape
    'merr_stress',0,
    0,'merr_quality'),
  DRIFT = c(  #temporal dependence matrix
    'a11','a12',
    'a21','a22'),
  DIFFUSION = c( #system noise. sd's on diagonal, unconstrained correlations on lower triangle
    'diff11',0,
    'diff21','diff22'),
  CINT = 0 #single values are repeated as many times as necessary (here, twice, one for each process)
)
  


badfit <- ctStanFit(datalong = dat,ctstanmodel = badmodel, cores=6)

ctKalman(badfit,kalmanvec=c('y','yprior'),plot=TRUE, subjects=1:2) #forward predictions
ctKalman(badfit,kalmanvec=c('y','ysmooth'),plot=TRUE, subjects=1:2) #best estimates conditional on all data

ctStanDiscretePars(badfit,plot=T)

```

**Comment by rifatzahan** [comment]
Thanks. Yes, it worked. Much appreciated your help.

**Reply by rifatzahan** [reply]
Hi Charles,

Thanks again for spending your valuable time on answering our questions. 

I in fact tried the above code for our datasets, which has four manifest variables. What me and my supervisor, are really wondering that is there a way to assess the impact of the time of the day on the whole dynamic system? Our data was collected three times a day everyday during the study period on a smartphone based platform called, Ethica. If it was any other type of regression analysis, I would have used a separate covariate, that has three labels to indicate the time of the day. I did not see in CTSEM (please correct me if I am wrong) any categorical variable used as manifest variable. 

Do you want to comment on this? Because we collect data from Ethica platform quite frequently for different projects in our lab. So time of the day comes up for such repeated measured data that we collect over the study period. 

Thanks in advance.

**Comment by cdriveraus** [comment]
You could make every parameter of the system dependent on time of day in the way you ask, and it might offer interesting insight or some kind of reasonable approximation, but these things get tricky when you are thinking about continuous processes through time -- if at 10:30am somebodies temporal effect parameter goes higher, what should be the assumption? Should we assume that the parameter only changed at that moment, and up until then it had the value assigned to the previous time of day covariate, say at 9pm? Do you expect some kind of linear or other interpolation between your time of day observations? The point is that parameters in a continuous time model are relevant at every time of the day, but with your desired covariate approach you are only saying how the parameters are at the few specific times of day you happen to observe. A more sophisticated approach would be to model some kind of extra time of day process, possibly as an oscillator, and then make your parameters dependent on the latent state. This would quickly get complicated but you can see an example of how to begin in the nonlinear section of the manual via `ctDocs()` 

The approach you were asking about is possible by including your times of day as distinct time dependent predictors, so e.g. a 'morning' column coded as 1 if it's morning and 0 otherwise, then making the parameters dependent on morning via something like:
```
DRIFT='baseDRIFT + morningDRIFT * tdpreds[rowi,1] + nightDRIFT * tdpreds[rowi,2]'
```

where the 1 and 2 refer to 1st and 2nd time dependent predictors

This all seems pretty complicated I assume, and that's kind of true, but if you want a serious interpretation of time of day moderation of dynamics I don't think there are simpler options. If you just want to account for time of day effects in the average levels of the process (and not the parameters), including some kind of oscillating process would be one option.
